### PR TITLE
Avoid leaking lease objects

### LINKED ISSUE #109777: Kubelet creates leases without ownerReference set if node doesn't exist
### What happened?

* Some external controller deleted node object while kubelet was alive
* Due to ignoring errors in kubelet's lease logic, it created lease without ownerReference set

### What did you expect to happen?

* Kubelet doesn't create lease without ownerReference set in any case, even pathological one

### How can we reproduce it (as minimally and precisely as possible)?

* Some external controller deletes node object
* Garbage collector properly deletes associated lease object
* Kubelet tries to update object and it fails on Conflict (saying that UID doesn't match): 
  https://github.com/kubernetes/kubernetes/blob/b1e130fe83156783153538b6d79821c2fdaa85bb/staging/src/k8s.io/component-helpers/apimachinery/lease/controller.go#L179-L198
* We go to `apierrors.IsConflict(err)` branch, call backoffEnsureLease and assign result to base, but backoffEnsureLease may return potentially invalid object if ensureLease returned error (which happens when newLeasePostProcessFunc returns error, e.g. when node doesn't exist).
* We `continue`
* The next loop calls c.newLease on the incorrect base and ignores error in line 181 (so the second attempt to call newLeasePostProcessFunc which also fails with an error) 
* We do UPDATE with incomplete (without effect of newLeasePostProcessFunc) object
* Given that leases allow CreateOnUpdate a new object, without ownerReference is created: https://github.com/kubernetes/kubernetes/blob/b1e130fe83156783153538b6d79821c2fdaa85bb/pkg/registry/coordination/lease/strategy.go#L65-L68

### Anything else we need to know?

_No response_

### Kubernetes version

<details>
1.21.10
</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by uesyn**
#### What type of PR is this?

/kind bug
/sig node

#### What this PR does / why we need it:

If a Node object is deleted before the kubelet is stopped, the lease controller in kubelet creates lease objects which has no ownerReference field. If the parent node is not re-registered, the lease object is leaked because its ownerReference field have not been updated forever.
This PR fixes it.

Lease has the special update endpoint which can only update but also create the new lease objects if it is not found.
In addition, when the lease controller get a conflict error in updating a lease object, it does the complicated operations other than get the latest lease object.
These behavior cause to create lease objects which has no ownerReference, and it is leaked.
For more information, see https://github.com/kubernetes/kubernetes/issues/109777#issuecomment-1168123849 and https://github.com/kubernetes/kubernetes/issues/109777#issue-1225134115.
This PR makes the lease controller not to create any new lease objects while updating not to leak the lease objects.

#### Which issue(s) this PR fixes:

Fixes #109777

#### Does this PR introduce a user-facing change?

```release-note
Fix leaking lease objects when a node is deleted before the kubelet is terminated.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by uesyn** [comment]
Passed the test.
I think the reason why unit-test failed is related to https://github.com/kubernetes/kubernetes/issues/110854.

**Comment by uesyn** [comment]
ping @caesarxuchao
 @jpbetz

**Comment by uesyn** [comment]
ping @caesarxuchao @jpbetz

**Comment by matthyx** [comment]
/remove-lifecycle stale
/lgtm
thanks @uesyn !

File: staging/src/k8s.io/component-helpers/apimachinery/lease/controller.go [hunk 1]
```diff
@@ -115,7 +115,9 @@ func (c *controller) sync() {
 	}
 
 	lease, created := c.backoffEnsureLease()
-	c.latestLease = lease
+	if lease != nil {
+		c.latestLease = lease
+	}
```
Comment (cartermckinnon) [inline staging/src/k8s.io/component-helpers/apimachinery/lease/controller.go:120]: This doesn't seem like the right approach.

`backoffEnsureLease` will return a `nil` lease if none exists in the API **and** `newLease` returns an error (such as when the `Node` doesn't exist): https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-helpers/apimachinery/lease/controller.go#L168

When that happens, if we *don't* set `c.latestLease` to `nil`, we'll enter the above `if` every time this function is called. `retryUpdateLease` will never succeed with your change below. We'll *then* call `ensureLease` here, with a chance of creating a new lease and stabilizing the state.

I think the approach in #119661 more directly addresses the issue, which is errors from `newLease` being ignored.

**Review by cartermckinnon — CHANGES_REQUESTED** [review]

**Pull request closed without merging**